java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return union(this, other);
+    }
+
+    public static TaintSource union(TaintSource left, TaintSource right) {
+      if (!left.canBeTainted()) {
+        return right;
+      }
+
+      if (!right.canBeTainted()) {
+        return left;
+      }
+
+      if (left.equals(right)) {
+        return left;
+      }
+
+      return new IndirectTaintSource(left, right);
+    }
+
+    @Override
+    public String toString() {
+      return "left: " + left.toString() + ", right = " + right.toString();
+    }
+
+    @Override
+    public int hashCode() {
+      return left.hashCode() + right.hashCode();
+    }
+
+    // TODO equals is not symmetrical
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      IndirectTaintSource other = (IndirectTaintSource) obj;
+      return left.equals(other.left) &&
+        right.equals(other.right);
+    }
+
+  }
+
+  private static class TaintProgramState {
+
+    private final ScannedFile src;
+    private final Map<Symbol, TaintSource> taintSources = new HashMap<>();
+    private final Set<DirectTaintSource> methodCalls = new HashSet<>();
+
+    public TaintProgramState(ScannedFile src) {
+      this.src = src;
+    }
+
+    private TaintProgramState(TaintProgramState other) {
+      this.src = other.src;
+      this.taintSources.putAll(other.taintSources);
+      this.methodCalls.addAll(other.methodCalls);
+    }
+
+    public TaintProgramState dup() {
+      return new TaintProgramState(this);
+    }
+
+    public SemanticModel semantic() {
+      return src.semantic();
+    }
+
+    private static String fullyQualify(Symbol s) {
+      if (s.isVariableSymbol()) {
+        Symbol owner = s.owner();
+        String ownerSignature;
+        if (owner.isTypeSymbol()) {
+          // field
+          ownerSignature = ((JavaSymbol.TypeJavaSymbol)owner).getFullyQualifiedName();
+        } else if (owner.isMethodSymbol()) {
+          // parameter
+          ownerSignature = ((JavaSymbol.MethodJavaSymbol)s.owner()).completeSignature();
+        } else {
+          throw new IllegalArgumentException();
+        }
+
+        return ownerSignature + "#" + s.name();
+      } else if (s.isMethodSymbol()) {
+        return ((JavaSymbol.MethodJavaSymbol)s).completeSignature();
+      } else {
+        throw new IllegalArgumentException();
+      }
+    }
+
+    public static boolean isString(Type type) {
+      return type.is("java.lang.String");
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol) {
+      return taintSourceFor(symbol, Collections.emptyList());
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol, List<TaintSource> arguments) {
+      if (symbol.isMethodSymbol()) {
+        MethodSymbol method = (MethodSymbol)symbol;
+
+        if (arguments.stream().anyMatch(ts -> ts.canBeTainted())) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612)